### PR TITLE
Display uploaded file name in the document title (#244)

### PR DESCRIPTION
**Pull request by pewstiepoll**
Hello!
This is the first time I contribute to open-source.
First I decided to divide an issue into smaller pieces:
1. Display filename in the title
2. Display filename in UI

I've already implemented the first part and would really like to hear feedback from you!

### CODE REVIEW

**Comment by jakearchibald** [comment]
@pewstiepoll honoured that this project is your first OS contribution.

I'm pretty happy with using the document title for this. We can figure out where else to display it later. What do you think @developit @kosamari @surma?

File: src/components/App/index.tsx [hunk 1]
```diff
@@ -60,8 +62,13 @@ export default class App extends Component<Props, State> {
       this.componentDidUpdate = (props, state) => {
         if (oldCDU) oldCDU.call(this, props, state);
         window.STATE = this.state;
+
+        // change title once component updated
+        this.changeTitle();
```
Comment (jakearchibald) [inline src/components/App/index.tsx]: This is inside `if (process.env.NODE_ENV === 'development') {` so it won't work in production.

File: src/components/App/index.tsx [hunk 2]
```diff
@@ -87,6 +94,19 @@ export default class App extends Component<Props, State> {
...
   }
 
+  @bind
+  private changeTitle() {
+    // Add the filename to the document title
+    const { file } = this.state;
+
+    // If there's a file - update title
+    if (file && file.name) {
+      document.title = `${this.defaultTitle} • ${file.name}`;
```
Comment (jakearchibald) [inline src/components/App/index.tsx]: I think we can put the file name first. Eg, https://www.bbc.co.uk/news/uk-politics-46269757 - the headline is in the title before "BBC news".

**Comment by pewstiepoll** [comment]
@jakearchibald thanks for quick reply!
About moving code to `compress/index.tsx` : in this case it will make Compress component perform the side effect (e.g. document changes) and from that perspective it would make it harder to debug in future. But from your presentation on Google Dev Summit I can assume that squoosh is about "first interaction" therefore moving things like this from init bundle makes sense. Just wanted to know your opinion about such trade-off.

About other comments: I agree in both cases and will fix it in my next changes.

**Comment by jakearchibald** [comment]
> in this case it will make Compress component perform the side effect

I see your point, but given it's *almost* the top level component, I'm fine with it. It also keeps the bulk of the "handle new image" code in one place.

**Comment by pewstiepoll** [comment]
@jakearchibald hi!
Just updated my changes corresponding to your review.
Also trying to figure out where to put filename in UI. Any suggestions?

**Comment by surma** [comment]
Why don't we scope this PR to document title and do a UI tweak separately? Wdyt?

**Comment by pewstiepoll** [comment]
@surma I think it makes sense. Just wonder if it's ok to have such a tiny changes in PR.

**Comment by surma** [comment]
No change is too small for its own PR. Makes it much easier to review and to reason about :)

File: src/components/compress/index.tsx [hunk 3]
```diff
@@ -162,6 +162,8 @@ const resultTitles = ['Top', 'Bottom'];
 const buttonPositions =
   ['download-left', 'download-right'] as ('download-left' | 'download-right')[];
 
+const originDocumentTitle = document.title;
```
Comment (jakearchibald) [inline src/components/compress/index.tsx]: This this a typo? As in, should it be `originalDocumentTitle`?

File: src/components/compress/index.tsx [hunk 4]
```diff
@@ -226,12 +228,23 @@ export default class Compress extends Component<Props, State> {
     });
   }
 
+  private updateDocumentTitle(filename : string = '') {
```
Comment (jakearchibald) [inline src/components/compress/index.tsx]: No space before the `:`, so `filename: string = ''`. Sorry, I thought our linter would catch this.

Comment (jakearchibald) [reply src/components/compress/index.tsx]: Also, when is filename empty?

File: src/components/compress/index.tsx [hunk 5]
```diff
@@ -226,12 +228,23 @@ export default class Compress extends Component<Props, State> {
...
+    const titleWithFilename = `${filename} - ${originDocumentTitle}`;
+    const newTitle : string = filename ? titleWithFilename : originDocumentTitle;
```
Comment (jakearchibald) [inline src/components/compress/index.tsx]: It seems odd to create `titleWithFilename` and then discard it. So:

```ts
document.title = filename ? `${filename} - ${originDocumentTitle}` : originDocumentTitle;
```

File: src/components/compress/index.tsx (hunk 4 above)
Comment (jakearchibald) [reply src/components/compress/index.tsx]: Ohh I see, it's on unmount. Gotcha.

File: src/components/compress/index.tsx [hunk 6]
```diff
@@ -328,7 +341,9 @@ export default class Compress extends Component<Props, State> {
         });
       }
 
-      this.setState(newState);
+      this.setState(newState, () => {
+        this.updateDocumentTitle(file.name);
```
Comment (jakearchibald) [inline src/components/compress/index.tsx]: I don't think we need to use the callback here, right? We can just call `updateDocumentTitle`.

Comment (pewstiepoll) [reply src/components/compress/index.tsx]: I thought we might want to display filename only after the file is loaded into the state, but if there's no reason - I'll change it.

**Comment by pewstiepoll** [comment]
@jakearchibald done.
Can't wait to see my changes in prod ^_^

File: src/components/compress/index.tsx (hunk 6 above)
Comment (jakearchibald) [reply src/components/compress/index.tsx]: You'll always see the result of `setState` before the next paint (unless I'm missing something), so without the callback is fine IMO.

**Comment by jakearchibald** [comment]
@pewstiepoll I'm really happy with this, thanks for making it happen & putting up with the review process.

**Comment by pewstiepoll** [comment]
@jakearchibald I’m so proud to make my first steps into OS under your guidance! Thank you! :)

**Comment by jakearchibald** [comment]
Your changes are now live 🎉

**Comment by pewstiepoll** [comment]
@jakearchbald, hooray!
Should I delete the branch now?

**Comment by jakearchibald** [comment]
You're welcome to (it's your branch). We don't need it anymore.

**Pull request merged**